modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/trait/DistributionTrait.java

```diff
@@ -34,73 +34,89 @@
         Ordering.<Integer>natural().lexicographical();
 
-    private RelDistribution.Type type;
+    /** */
+    private DistributionFunction function;
+
+    /** */
     private ImmutableIntList keys;
-    private DestinationFunctionFactory functionFactory;
 
-    public DistributionTrait() {
-    }
+    /**
+     * @param function Distribution function.
+     */
+    DistributionTrait(DistributionFunction function) {
+        assert function.type() != HASH_DISTRIBUTED;
 
-    public DistributionTrait(RelDistribution.Type type, ImmutableIntList keys, DestinationFunctionFactory functionFactory) {
-        if (type == RANGE_DISTRIBUTED || type == ROUND_ROBIN_DISTRIBUTED)
-            throw new IllegalArgumentException("Distribution type " + type + " is unsupported.");
+        this.function = function;
 
-        this.type = type;
+        keys = ImmutableIntList.of();
+    }
+
+    /**
+     * @param keys Distribution keys.
+     * @param function Distribution function.
+     */
+    DistributionTrait(ImmutableIntList keys, DistributionFunction function) {
         this.keys = keys;
-        this.functionFactory = functionFactory;
+        this.function = function;
     }
 
+    /** {@inheritDoc} */
     @Override public RelDistribution.Type getType() {
-        return type;
+        return function.type();
     }
 
-    @Override public DestinationFunctionFactory destinationFunctionFactory() {
-        return functionFactory;
+    /** {@inheritDoc} */
+    @Override public DistributionFunction function() {
+        return function;
     }
 
+    /** {@inheritDoc} */
     @Override public ImmutableIntList getKeys() {
         return keys;
     }
 
+    /** {@inheritDoc} */
     @Override public void register(RelOptPlanner planner) {}
 
+    /** {@inheritDoc} */
     @Override public boolean equals(Object o) {
         if (this == o)
             return true;
 
         if (o instanceof DistributionTrait) {
             DistributionTrait that = (DistributionTrait) o;
 
-            return type == that.type
-                && Objects.equals(keys, that.keys)
-                && Objects.equals(functionFactory.key(), that.functionFactory.key());
+            return Objects.equals(function, that.function) && Objects.equals(keys, that.keys);
         }
 
         return false;
     }
 
+    /** {@inheritDoc} */
     @Override public int hashCode() {
-        return Objects.hash(type, keys);
+        return Objects.hash(function, keys);
     }
 
+    /** {@inheritDoc} */
     @Override public String toString() {
-        return type + (type == Type.HASH_DISTRIBUTED ? "[" + functionFactory.key() + "]" + keys : "");
+        return function.name() + (function.type() == HASH_DISTRIBUTED ? keys : "");
     }
 
+    /** {@inheritDoc} */
     @Override public DistributionTraitDef getTraitDef() {
         return DistributionTraitDef.INSTANCE;
     }
 
+    /** {@inheritDoc} */
     @Override public boolean satisfies(RelTrait trait) {
```

Comment:
Don't forget to add BROADCAST.satisfies(SINGLETON)

Reply:
I'll try)